ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/PeerSimulator/NodeLifecycle.hs

```diff
@@ -0,0 +1,208 @@
+import           Test.Consensus.PeerSimulator.Resources
+import           Test.Consensus.PeerSimulator.StateView
+import           Test.Consensus.PeerSimulator.Trace
+import           Test.Consensus.PointSchedule.Peers (PeerId)
+import           Test.Util.ChainDB
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.TestBlock (TestBlock, testInitExtLedger)
+
+-- | Resources used for a single live interval of the node, constructed when the
+-- node is started.
+-- When the node is shut down, 'lnCopyToImmDb' is used to persist the immutable
+-- chain.
+data LiveNode blk m = LiveNode {
+    lnChainDb          :: ChainDB m blk
+  , lnStateViewTracers :: StateViewTracers blk m
+  , lnStateTracer      :: Tracer m ()
+  , lnCopyToImmDb      :: m (WithOrigin SlotNo)
+
+    -- | The set of peers that should be started.
+    -- Based on the simulation results at node shutdown, disconnected peers are
+    -- removed for the next live interval.
+  , lnPeers            :: Set PeerId
+  }
+
+-- | Result of a node shutdown at the end of a live interval.
+data LiveIntervalResult blk = LiveIntervalResult {
+    -- | Used to initialize the 'StateViewTracers' of the next run to preserve
+    -- earlier disconnections for the final result.
+    lirPeerResults :: [PeerSimulatorResult blk]
+
+    -- | The remaining peers, computed by removing all peers present in
+    -- 'lrPeerResults' from the current state in 'lnPeers'.
+  , lirActive      :: Set PeerId
+  }
+
+-- | Resources used by the handlers in 'NodeLifecycle' to shut down running
+-- components, construct tracers used for single intervals, and reset and
+-- persist state.
+data LiveResources blk m = LiveResources {
+    lrRegistry :: ResourceRegistry m
+  , lrPeerSim  :: PeerSimulatorResources m blk
+  , lrTracer   :: Tracer m (TraceEvent blk)
+  , lrSTracer  :: ChainDB m blk -> m (Tracer m ())
+  , lrConfig   :: TopLevelConfig blk
+
+    -- | The chain DB state consists of several transient parts and the
+    -- immutable DB's virtual file system.
+    -- After 'lnCopyToImmDb' was executed, the latter will contain the final
+    -- state of an interval.
+    -- The rest is reset when the chain DB is recreated.
+  , lrCdb      :: NodeDBs (StrictTVar m MockFS)
+
+    -- | The LoE fragment must be reset for each live interval.
+  , lrLoEVar   :: LoE (StrictTVar m (AnchoredFragment (Header blk)))
+  }
+
+data LiveInterval blk m = LiveInterval {
+    liResources :: LiveResources blk m
+  , liResult    :: LiveIntervalResult blk
+  , liNode      :: LiveNode blk m
+  }
+
+-- | Handlers for starting the node and shutting it down for each live interval,
+-- using the state of the previous run.
+data NodeLifecycle blk m = NodeLifecycle {
+    -- | The minimum tick duration that triggers a node downtime.
+    -- If this is 'Nothing', downtimes are disabled.
+    nlMinDuration :: Maybe DiffTime
+  , nlStart       :: LiveIntervalResult blk -> m (LiveNode blk m)
```

Comment:
Is this handler also used for the very first start? With a dummy `LiveIntervalResult`, I assume; something like
```haskell
LiveIntervalResult {
  lirPeerResults = [],
  lirActive = ...set of all peers in the schedule...
}
```

Reply:
yes, updated the doc. ✅